service/src/main/java/org/cbioportal/service/impl/GeneServiceImpl.java

```diff
@@ -31,50 +31,138 @@
+        return geneRepository.getMetaGenes();
+    }
+
+    @Override
+    public Gene getGene(String geneId) throws GeneNotFoundException {
+
+        Gene gene;
+
+        if (isInteger(geneId)) {
+            gene = geneRepository.getGeneByEntrezGeneId(Integer.valueOf(geneId));
+        } else {
+            gene = geneRepository.getGeneByHugoGeneSymbol(geneId);
+        }
+
+        if (gene == null) {
+            throw new GeneNotFoundException(geneId);
+        }
+
+        gene.setChromosome(getChromosome(gene.getCytoband()));
+        return gene;
+    }
+
+    @Override
+    public List<String> getAliasesOfGene(String geneId) {
+
+        if (isInteger(geneId)) {
+            return geneRepository.getAliasesOfGeneByEntrezGeneId(Integer.valueOf(geneId));
+        } else {
+            return geneRepository.getAliasesOfGeneByHugoGeneSymbol(geneId);
+        }
+    }
+
+    @Override
+    public List<Gene> fetchGenes(List<String> geneIds, String projection) {
+
+        List<Integer> entrezGeneIds = new ArrayList<>();
+        List<String> hugoGeneSymbols = new ArrayList<>();
+
+        splitIdsByType(geneIds, entrezGeneIds, hugoGeneSymbols);
+
+        List<Gene> geneList = geneRepository.fetchGenesByEntrezGeneIds(entrezGeneIds, projection);
+        geneList.addAll(geneRepository.fetchGenesByHugoGeneSymbols(hugoGeneSymbols, projection));
+
+        for (Gene gene : geneList) {
+            gene.setChromosome(getChromosome(gene.getCytoband()));
+        }
+
+        return geneList;
+    }
+
+
+
+    @Override
+    public BaseMeta fetchMetaGenes(List<String> geneIds) {
+
+        List<Integer> entrezGeneIds = new ArrayList<>();
+        List<String> hugoGeneSymbols = new ArrayList<>();
+
+        splitIdsByType(geneIds, entrezGeneIds, hugoGeneSymbols);
+
+        BaseMeta baseMeta = new BaseMeta();
+        baseMeta.setTotalCount(geneRepository.fetchMetaGenesByEntrezGeneIds(entrezGeneIds).getTotalCount() +
+                geneRepository.fetchMetaGenesByHugoGeneSymbols(hugoGeneSymbols).getTotalCount());
+
+        return baseMeta;
+    }
+
+    private void splitIdsByType(List<String> geneIds, List<Integer> entrezGeneIds, List<String> hugoGeneSymbols) {
+
+        for (String geneId : geneIds) {
+            if (isInteger(geneId)) {
+                entrezGeneIds.add(Integer.valueOf(geneId));
+            } else {
+                hugoGeneSymbols.add(geneId);
+            }
+        }
+    }
+
+    private boolean isInteger(String geneId) {
+        return geneId.matches("^-?\\d+$");
+    }
+
+    private String getChromosome(String cytoband) {
+
+        if (cytoband == null) {
             return null;
         }
-        if (gene.getCytoband().toUpperCase().startsWith("X")) {
+
+        cytoband = cytoband.toUpperCase();
+        if (cytoband.startsWith("X")) {
             return "X";
-        }
-        if (gene.getCytoband().toUpperCase().startsWith("Y")) {
+        } else if (cytoband.startsWith("Y")) {
```

Comment:
I'll point out again that some Cytoband entries contain a combination of X and Y .. for genes located on both X and Y. We may want to model this properly .. although for now I think all actual cases start with X and nearly all individuals do have an X chromosome so in a simplified model this is pretty close to "true". See https://github.com/cBioPortal/cbioportal/issues/1699